Add mode control for Modbus climate entities

avishorp opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add support for more features for a climate entity using Modbus platform.

This PR adds the ability to configure registers to control and monitor a climate
entity's state (heat, cool, on/off, etc.)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
This PR adds two optional configuration entries to a Modbus climate entity. The first one is
a mode register, which is used to determine the operating mode of an HVAC. An additional
set of configuration items map between the possible states and the values read/written to this
register.

The second register is an On/Off register, which enables turning a climate entity on or off (in cases
where this functionality is not provided by the mode register).

The code was tested against a Samsung air condition system, and from the documentation also seem
to be compatible with systems that are controlled by [Intesis](https://www.intesis.com/) interface system which
is available for many AC brands.

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/23231

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

avishorp committed:
Add support for Modbus HVAC control registers

avishorp committed:
Reformat some files

avishorp committed:
Fix incorrect attribute variable name

avishorp committed:
Add tests for Modbus climate device

avishorp committed:
Fix linting errors

avishorp committed:
Fix some more linting problems

avishorp committed:
Merge branch 'home-assistant:dev' into dev

avishorp committed:
Merge branch 'dev' of github.com:avishorp/hass-core into dev

homeassistant commented:
Hi @avishorp, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#73906).

Once you do that we will be able to review and accept this pull request.

Thanks!

janiversen commented:
The idea quite good I will take a deeper look later.

janiversen commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -216,6 +225,15 @@
             vol.Optional(CONF_MIN_TEMP, default=5): cv.positive_int,
             vol.Optional(CONF_STEP, default=0.5): vol.Coerce(float),
             vol.Optional(CONF_TEMPERATURE_UNIT, default=DEFAULT_TEMP_UNIT): cv.string,
+            vol.Optional(CONF_HVAC_MODE_REGISTER): cv.positive_int,
```
It is better to have CONF_HVAC_MODE as optional, and the below it the different modes.

avishorp replied:
This was my initial thought also, but I tried to follow the convention used in MQTT configuration, where both the name of the register (topic in MQTT) and the set of possible values are defined at the same level. If you insist, I can change it but I think it might confuse users. This is also the reason I needed extra constant for each possible mode.

janiversen replied:
Yes I insist, because that is what Modbus does with other parameters of similar type (e.g. 'verify').

janiversen commented on homeassistant/components/modbus/const.py:
```diff
@@ -53,6 +53,15 @@
 CONF_SWAP_WORD = "word"
 CONF_SWAP_WORD_BYTE = "word_byte"
 CONF_TARGET_TEMP = "target_temp_register"
+CONF_HVAC_MODE_REGISTER = "hvac_mode_register"
```
We should max. define HVAC_MODE, and use the consts already defined in components/climate for each entry.

avishorp replied:
Done.

janiversen reviewed (approved):
Please follow the suggestions made.

Once you have made the configuration changes, you need to document it.

janiversen reviewed (changes requested):
Sorry hit the wrong button.

avishorp commented:
Documentation is on the way, will submit a PR soon.

avishorp committed:
Change the climate mode configuration scheme

avishorp committed:
Format the source

avishorp committed:
Use HVACMode instead of the deprecated HVAC_MODE_* constants

avishorp committed:
Migrate one more file to the new constants

avishorp committed:
Fix bad configuration in test

avishorp committed:
Merge branch 'dev' into dev

avishorp committed:
Merge branch 'dev' into dev

janiversen commented:
I have added some review comments to the doc. please change the code accordingly.

avishorp commented:
I made the requested changes, do I need to do anything else to get this PR merged?

avishorp committed:
Combine Modbus read functions and remove duplicate code

avishorp committed:
Change test to cover another case

avishorp committed:
Merge branch 'dev' of github.com:avishorp/hass-core into dev

avishorp committed:
Fix linting and formatting

avishorp committed:
Remove unused import

avishorp committed:
Merge branch 'dev' into dev

jiagangxu commented:
@avishorp, is it possible to also add fan mode control to modbus climate? I know the modbus component updates from time to time, and there are quite a lot users customize modbus climate to add more registers to suit for their devices. It will save the trouble for end users to update customized components along with modbus update. Thanks for your support!

avishorp commented:
I will happily do that, but currently the PR is stuck, I don't know why

‫בתאריך יום א׳, 28 באוג׳ 2022 ב-12:16 מאת ‪Jiagang Xu‬‏ <‪
***@***.***‬‏>:‬

> @avishorp <https://github.com/avishorp>, is it possible to also add fan
> mode control to modbus climate? I know the modbus component updates from
> time to time, and there are quite a lot users customize modbus climate to
> add more registers to suit for their devices. It will save the trouble for
> end users to update customized components along with modbus update. Thanks
> for your support!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/73906#issuecomment-1229413824>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJCMMTA562AMO2UHJXNNX3V3MU5PANCNFSM5ZUQZXHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

janiversen commented:
Waiting for review, and if you look at CI you have a doc problem. I will review it once CI is green.

avishorp commented:
I have made the required doc changes (also waiting for review), how do I
link it to the feature PR?

‫בתאריך יום ב׳, 29 באוג׳ 2022 ב-14:27 מאת ‪jan iversen‬‏ <‪
***@***.***‬‏>:‬

> Waiting for review, and if you look at CI you have a doc problem. I will
> review it once CI is green.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/73906#issuecomment-1230156539>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJCMMVLPWVNAJZ4A6LVPD3V3SNAPANCNFSM5ZUQZXHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

janiversen commented:
See the PR text (Proposed change etc.), seems you have modified the template, normally it contains:
```
## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
```

Btw. you need to link this PR in the documentation PR as well.

janiversen committed:
Merge branch 'dev' into dev

jiagangxu commented on homeassistant/components/modbus/climate.py:
```diff
@@ -87,6 +89,33 @@ def __init__(
         self._attr_target_temperature_step = config[CONF_TARGET_TEMP]
```
Though this part remains the same, there is a question.
line 89 and line 90 have the same lvalue. Maybe line 89 should be self._attr_target_temperature = config[CONF_TARGET_TEMP]?

avishorp committed:
Merge branch 'dev' into dev

avishorp committed:
Merge branch 'dev' into dev

avishorp commented:
I've changed the text and added the missing lines with a link, still
getting docs-missing.

‫בתאריך יום ב׳, 29 באוג׳ 2022 ב-14:41 מאת ‪jan iversen‬‏ <‪
***@***.***‬‏>:‬

> See the PR text (Proposed change etc.), seems you have modified the
> template, normally it contains:
>
> ## Additional information
> <!--
>   Details are important, and help maintainers processing your PR.
>   Please be sure to fill out additional details, if applicable.
> -->
>
> - This PR fixes or closes issue: fixes #
> - This PR is related to issue:
> - Link to documentation pull request:
>
> Btw. you need to link this PR in the documentation PR as well.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/73906#issuecomment-1230171201>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJCMMSSYO33YC66SEUUTQ3V3SOW7ANCNFSM5ZUQZXHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

avishorp committed:
Fix import sort order

janiversen reviewed (approved):
LGTM.

frenck commented:
@avishorp There is a merge conflict, could you take a look?

Thanks! 👍 

../Frenck

avishorp committed:
Merge branch 'dev' into dev

avishorp commented:
Fixed the merge conflict.

stillartem commented:
@frenck Any updates or concerns? does production see soon this MR ?

janiversen commented:
There is a pylint error prohibiting merge, please fix.

stillartem commented:
@avishorp  you did a brilliant job, could you make one more fix to complete all checks, please

avishorp committed:
Fixed root import linting error

avishorp committed:
Run isort

avishorp commented:
> 

Fixed
